plugin.video.psvue/resources/lib/guideservice.py

```diff
@@ -0,0 +1,179 @@
+import math
+from globals import *
+from database import Database
+
+
+class BuildGuide(threading.Thread):
+    guide_days = int(ADDON.getSetting('epg_days'))
+    guide_thread_1 = None
+    guide_thread_2 = None
+    guide_thread_3 = None
+    guide_thread_4 = None
+    keep_running = True
+    monitor = xbmc.Monitor()
+    update_interval = int(ADDON.getSetting('epg_interval')) * 3600
+    update_on_start = ADDON.getSetting('epg_on_start')
+
+    def __init__(self):
+        threading.Thread.__init__(self)
+        self.db = Database()
+
+    def run(self):
+        xbmc.log('BuildGuide: Thread starting....')
+
+        first_time_thru = True
+        if self.update_on_start == 'false':
+            first_time_thru = False
+            self.monitor.waitForAbort(self.update_interval)
+
+        while not self.monitor.abortRequested():
+            xbmc.log('BuildGuide: Looping through guide days....')
+            # Build main guide longer w/ less info
+            self.guide_thread_1 = threading.Thread(name='GuideThread 1', target=self.long_guide())
+
+            while self.guide_thread_1.isAlive():
+                xbmc.log('BuildGuide: Main guide thread active, waiting for finish')
+
+            # Build short guide with more info
+            channel_ids = []
+            for channel_id, title, logo in self.db.get_db_channels():
+                channel_ids.append(channel_id)
+
+            third = int(math.ceil(len(channel_ids) / 3))
+            self.guide_thread_2 = threading.Thread(name='GuideThread 2',
+                                                   target=self.short_guide(channel_ids[:third]))
+            self.guide_thread_3 = threading.Thread(name='GuideThread 3',
+                                                   target=self.short_guide(channel_ids[third:third + third]))
+            self.guide_thread_4 = threading.Thread(name='GuideThread 4',
+                                                   target=self.short_guide(channel_ids[third + third:]))
+
+            xbmc.log('BuildGuide: before loop')
+            thread_alive = True
+            while thread_alive:
+                thread_alive = False
+                if self.guide_thread_2.isAlive() or self.guide_thread_3.isAlive() or self.guide_thread_4.isAlive():
+                    thread_alive = True
+                    xbmc.log('BuildGuide: Active threads remain, waiting for finish')
+
+                if self.monitor.waitForAbort(5):
+                    break
+
+            xbmc.log('BuildGuide: after loop')
+            self.db.clean_db_epg()
+            self.db.build_epg_xml()
+            if first_time_thru:
+                xbmc.log('BuildGuide: First time thru, toggling IPTV restart')
+                pvr_toggle_off = '{"jsonrpc": "2.0", "method": "Addons.SetAddonEnabled", ' \
```

Comment:
Not sure if this is allowed without first prompting the user? @enen92 input?

Reply:
Yes please add a yes/no before the call